File: java/concurrent/src/test/java/cn/wjybxx/sequential/UniScheduleOrderTest.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.sequential;

import cn.wjybxx.base.ThreadUtils;
import cn.wjybxx.base.time.TimeProviders;
import cn.wjybxx.concurrent.Counter;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

/**
 * @author wjybxx
 * date - 2024/4/3
 */
public class UniScheduleOrderTest {

    private Counter counter;
    private UniScheduledExecutor consumer;

    @BeforeEach
    void setUp() {
        counter = new Counter();
        consumer = new DefaultUniScheduledExecutor(TimeProviders.systemMillisProvider());
    }

    /** 测试initDelay相同时，任务是否按照按提交顺序执行 */
    @Test
    void testScheduleOrder() {
        final ThreadLocalRandom random = ThreadLocalRandom.current();
        final TimeUnit milliseconds = TimeUnit.MILLISECONDS;
        for (int i = 0; i < 100; i++) {
            Runnable newTask = counter.newTask(1, i);
            switch (random.nextInt(3)) {
                case 1 -> consumer.scheduleWithFixedDelay(newTask, 100, 200, milliseconds);
                case 2 -> consumer.scheduleAtFixedRate(newTask, 100, 200, milliseconds);
                default -> consumer.schedule(newTask, 100, milliseconds);
            }
        }

        for (int i = 0; i < 100; i++) {
            consumer.update();
            ThreadUtils.sleepQuietly(1);
        }
        consumer.shutdownNow();

        Assertions.assertTrue(!counter.getSequenceMap().isEmpty(), "Counter.sequenceMap.size == 0");
        Assertions.assertTrue(counter.getErrorMsgList().isEmpty(), counter.getErrorMsgList()::toString);
    }

    /** 测试execute和schedule(0)的顺序 */
    @Test
    void testExecuteScheduleOrder() {
        final ThreadLocalRandom random = ThreadLocalRandom.current();
        final TimeUnit milliseconds = TimeUnit.MILLISECONDS;
        for (int i = 0; i < 100; i++) {
            Runnable newTask = counter.newTask(1, i);
            switch (random.nextInt(3)) {
                case 1 -> consumer.execute(newTask);
                case 2 -> consumer.submitAction(newTask);
                default -> consumer.schedule(newTask, 0, milliseconds);
            }
        }

        for (int i = 0; i < 100; i++) {
            consumer.update();
            ThreadUtils.sleepQuietly(1);
        }
        consumer.shutdownNow();

        Assertions.assertTrue(!counter.getSequenceMap().isEmpty(), "Counter.sequenceMap.size == 0");
        Assertions.assertTrue(counter.getErrorMsgList().isEmpty(), counter.getErrorMsgList()::toString);
    }
}

File: csharp/Wjybxx.Commons.Tests/src/Sequential/UniScheduleOrderTest.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Threading;
using Commons.Tests.Concurrent;
using NUnit.Framework;
using Wjybxx.Commons.Sequential;
using Wjybxx.Commons.Time;

namespace Commons.Tests.Sequential;

public class UniScheduleOrderTest
{
    private static Counter counter;
    private static IUniScheduledExecutor consumer;

    [SetUp]
    public void SetUp() {
        counter = new Counter();
        consumer = new DefaultUniScheduledExecutor(TimeProviders.SystemMillisProvider());
    }

    /// <summary>
    /// 测试initDelay相同时，任务是否按照按提交顺序执行
    /// </summary>
    [Test]
    public void TestScheduleOrder() {
        Random random = new Random();
        TimeSpan initDelay = TimeSpan.FromMilliseconds(100);
        TimeSpan period = TimeSpan.FromMilliseconds(200);
        for (int i = 0; i < 100; i++) {
            Action newTask = counter.NewTask(1, i);
            switch (random.Next(3)) {
                case 1: {
                    consumer.ScheduleWithFixedDelay(newTask, initDelay, period);
                    break;
                }
                case 2: {
                    consumer.ScheduleAtFixedRate(newTask, initDelay, period);
                    break;
                }
                default: {
                    consumer.ScheduleAction(newTask, initDelay);
                    break;
                }
            }
        }

        for (int i = 0; i < 100; i++) {
            consumer.Update();
            Thread.Sleep(1);
        }
        consumer.ShutdownNow();

        Assert.IsTrue(counter.sequenceMap.Count > 0, "counter.sequenceMap.Count > 0");
        Assert.IsTrue(counter.errorMsgList.Count == 0, "counter.errorMsgList.Count == 0");
    }

    /// <summary>
    /// 测试execute和schedule(0)的顺序
    /// </summary>
    [Test]
    public void TestExecuteScheduleOrder() {
        Random random = new Random();
        for (int i = 0; i < 100; i++) {
            Action newTask = counter.NewTask(1, i);
            switch (random.Next(3)) {
                case 1: {
                    consumer.Execute(newTask);
                    break;
                }
                case 2: {
                    consumer.SubmitAction(newTask);
                    break;
                }
                default: {
                    consumer.ScheduleAction(newTask, TimeSpan.Zero);
                    break;
                }
            }
        }

        for (int i = 0; i < 100; i++) {
            consumer.Update();
            Thread.Sleep(1);
        }
        consumer.ShutdownNow();

        Assert.IsTrue(counter.sequenceMap.Count > 0, "counter.sequenceMap.Count > 0");
        Assert.IsTrue(counter.errorMsgList.Count == 0, "counter.errorMsgList.Count == 0");
    }
}